File: algorithms/code/leetcode/lc110_balanced_binary_tree/lc110_balanced_binary_tree.py
# https://leetcode.com/problems/balanced-binary-tree/

from __future__ import annotations

import unittest


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: TreeNode | None = None,
        right: TreeNode | None = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __balancedHeight(self, root: TreeNode | None) -> int:
        if root is None:
            return 0
        left = self.__balancedHeight(root.left)
        if left == -1:
            return -1
        right = self.__balancedHeight(root.right)
        if right == -1:
            return -1
        if abs(left - right) > 1:
            return -1
        return 1 + max(left, right)

    def isBalanced(self, root: TreeNode | None) -> bool:
        return self.__balancedHeight(root) != -1


class TestCode(unittest.TestCase):
    def test_balanced(self) -> None:
        left = TreeNode(2, TreeNode(1, None, None), TreeNode(3, None, None))
        right = TreeNode(6, TreeNode(5, None, None), TreeNode(7, None, None))
        root = TreeNode(4, left, right)
        assert Solution().isBalanced(root)

    def test_imbalanced_right(self) -> None:
        right = TreeNode(6, TreeNode(5, None, None), TreeNode(7, None, None))
        root = TreeNode(4, None, right)
        assert not Solution().isBalanced(root)

    def test_imbalanced_left(self) -> None:
        left = TreeNode(2, TreeNode(1, None, None), TreeNode(3, None, None))
        root = TreeNode(4, None, left)
        assert not Solution().isBalanced(root)

    def test_imbalanced_deep_left(self) -> None:
        left = TreeNode(2, TreeNode(1, TreeNode(3, None, None), None), None)
        right = TreeNode(6, TreeNode(5, None, None), TreeNode(7, None, None))
        root = TreeNode(4, left, right)
        assert not Solution().isBalanced(root)

    def test_imbalanced_deep_right(self) -> None:
        left = TreeNode(2, TreeNode(1, None, None), TreeNode(3, None, None))
        right = TreeNode(6, TreeNode(5, None, TreeNode(7, None, None)), None)
        root = TreeNode(4, left, right)
        assert not Solution().isBalanced(root)

File: algorithms/code/leetcode/lc110_balanced_binary_tree/lc110_balanced_binary_tree.rs
// https://leetcode.com/problems/balanced-binary-tree/
// #easy

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<Self>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }
}

pub struct Solution;

impl Solution {
    fn balanced_height(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if let Some(root) = root {
            let root = root.borrow();
            let left = Self::balanced_height(root.left.clone());
            if left == -1 {
                return -1;
            }
            let right = Self::balanced_height(root.right.clone());
            if right == -1 {
                return -1;
            }
            if (left - right).abs() > 1 {
                -1
            } else {
                1 + left.max(right)
            }
        } else {
            0
        }
    }

    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::balanced_height(root) != -1
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_balanced() {
        let left = TreeNode::new(
            2,
            TreeNode::new(1, None, None),
            TreeNode::new(3, None, None),
        );
        let right = TreeNode::new(
            6,
            TreeNode::new(5, None, None),
            TreeNode::new(7, None, None),
        );
        let root = TreeNode::new(4, left, right);
        assert!(Solution::is_balanced(root));
    }

    #[test]
    fn test_imbalanced_right() {
        let right = TreeNode::new(
            6,
            TreeNode::new(5, None, None),
            TreeNode::new(7, None, None),
        );
        let root = TreeNode::new(4, None, right);
        assert!(!Solution::is_balanced(root));
    }

    #[test]
    fn test_imbalanced_left() {
        let left = TreeNode::new(
            2,
            TreeNode::new(1, None, None),
            TreeNode::new(3, None, None),
        );
        let root = TreeNode::new(4, left, None);
        assert!(!Solution::is_balanced(root));
    }

    #[test]
    fn test_imbalanced_deep_left() {
        let left = TreeNode::new(
            2,
            TreeNode::new(1, TreeNode::new(3, None, None), None),
            None,
        );
        let right = TreeNode::new(
            6,
            TreeNode::new(5, None, None),
            TreeNode::new(7, None, None),
        );
        let root = TreeNode::new(4, left, right);
        assert!(!Solution::is_balanced(root));
    }

    #[test]
    fn test_imbalanced_deep_right() {
        let left = TreeNode::new(
            2,
            TreeNode::new(1, None, None),
            TreeNode::new(3, None, None),
        );
        let right = TreeNode::new(
            6,
            TreeNode::new(5, None, TreeNode::new(7, None, None)),
            None,
        );
        let root = TreeNode::new(4, left, right);
        assert!(!Solution::is_balanced(root));
    }
}
